### PR TITLE
Front page flex-box

### LINKED ISSUE #583: Implementing a flex-box for front page course-sections
The course-items on the front page have a fixed height at the moment which might cause problems in case the course-name or course instance -name are very long. Or very short. 

Therefore it would be good to organize the items with a flexbox-container instead on the frontpage, so that the height of each item on the same course-row would grow if one needs extra space, and the grid would look clean and present all the course headlines clearly.

### PR DESCRIPTION
**Pull request by saarasat**
# Description

In order to better accommodate longer course and course instance -names, the bootstrap3 row-column-grid is removed from front page sections and flex-box is applied instead. This enables the course cards to grow vertically.

Four media-breakpoints have been applied, so that the width of the cards would look good with different screens.

The second commit 57e54963a is only to fix the indendation of the adverts, so that the changes made in the first commit can be more easily spotted.

Fixes #583

**Desktop width**

Screenshot from 2020-07-03 10-17-35

**Tablet width**

Screenshot from 2020-07-03 10-18-48

**Mobile width**

Screenshot from 2020-07-03 10-19-14

# Testing

- The appearance after changes has been tested on Chrome and Firefox and different screen widths.
- The appearance has also been tested with Lambdatest on Edge and IE. 
- It has also been tested for very long and very short course and course instance names
- Appearance has been tested when there are very few course instances and very many instances.
- The appearance has been tested also in case there is a site-wide advert present

# Have you updated the README or other relevant documentation?

Not relevant

# Is it [Done](https://wiki.aalto.fi/display/EDIT/Definition+of+Done)?

- [ ] I (developer) have created unit tests and the tests pass
- [ ] I (developer) have created functional tests (Selenium tests) if applicable
- [x] I (developer) have tested the changes manually
- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

### CODE REVIEW

File: assets/css/main.css [hunk 1]
```diff
@@ -325,10 +334,38 @@ html, body {
     border-color: #bce8f1;
 }
 
+@media (min-width: 1200px) and (max-width: 1800px) {
+    .frontpage .card {
+        width: 23%;
+    }
+}
+
+@media (min-width: 960px) and (max-width: 1200px) {
```
Comment (mtthwhggns) [inline assets/css/main.css:343]: I wonder if it would make sense if these matched the Bootstrap ones which affect our other layouts - https://getbootstrap.com/docs/3.4/css/#grid-media-queries

File: assets/css/main.css [hunk 2]
```diff
@@ -325,10 +334,38 @@ html, body {
...
+        width: 100%;
+    }
+}
+
 .frontpage a:hover {
     text-decoration: none;
 }
 
+.frontpage .card-clickable {
+    flex-grow: 1;
```
Comment (mtthwhggns) [inline assets/css/main.css:366]: What happens if this isn't applied?

File: course/templates/course/index.html [hunk 3]
```diff
@@ -16,7 +16,7 @@ <h1>{{ welcome_text|safe }}</h1>
 	<div class="frontpage-panel">
 		<h2 class="panel-heading">{% trans "My courses" %}</h2>
 		{% if my_instances %}
-			<div class="row frontpage">
+			<div class="frontpage frontpage-section">
```
Comment (mtthwhggns) [inline course/templates/course/index.html:19]: I wonder if [`<section>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section) would make more sense semantically here?

File: assets/css/main.css (hunk 1 above)
Comment (saarasat) [reply assets/css/main.css:343]: I thought about this too and actually tried those. I also tried to find information of if there are any newer, "standard" breakpoints, but they seemed to vary among different sources. The breakpoints I used looked better for these course-cards, so that is the reason I went for those. 

I guess this is about whether we want to standardize the breakpoints throughout, with a bit worse appearance on the front page or customize these here.

File: assets/css/main.css (hunk 2 above)
Comment (saarasat) [reply assets/css/main.css:366]: This makes sure that the footer will stick to the bottom of the card, and the clickable part is the one that grows, if there is need for the cards to vertically grow.

File: course/templates/course/index.html (hunk 3 above)
Comment (saarasat) [reply course/templates/course/index.html:19]: Good point, did not thought of that. Though the `<section>`-tag is not used elsewhere in the code... What do you think of using different tags even though the rest of the code is not using them? If it is ok, I could apply it after I hear I reviews and possible change-ideas of others.

File: assets/css/main.css (hunk 2 above)
Comment (mtthwhggns) [reply assets/css/main.css:366]: Ah, interesting! Thanks

File: course/templates/course/index.html (hunk 3 above)
Comment (raphendyr) [reply course/templates/course/index.html:19]: > What do you think of using different tags even though the rest of the code is not using them?

If they make the site better, feel free to make the site better :)

Comment (mtthwhggns) [reply course/templates/course/index.html:19]: That's fine!

**Comment by raphendyr** [comment]
If Aalto VDI has IE 11, this change could be checked there. I'm not sure how, but I think the front page should be usable to some extend with that browser.

File: course/templates/course/index.html (hunk 3 above)
Comment (Mankro) [reply course/templates/course/index.html:19]: Like Matthew has suggested earlier, there are small accessibility issues that can be fixed when someone is working on the relevant code. For example, fixing the semantic markup for sections fits here.

File: assets/css/main.css [hunk 4]
```diff
@@ -325,10 +334,38 @@ html, body {
...
+        width: 47%;
+    }
+}
+
+@media (max-width: 670px) {
+    .frontpage .card {
+        width: 100%;
+    }
+}
+
```
Comment (etanttila) [inline assets/css/main.css:360]: These min-widths and max-widths overlap. In border cases, it seems to use the latest applicable definition. This probably isn't a big problem but it might be better to use restrictions with a difference of a pixel (eg. `max-width: 1199px` and of another:` min-width: 1200` px (or 1200px and 1201px)).

File: locale/fi/LC_MESSAGES/django.po [hunk 5]
```diff
@@ -6,7 +6,7 @@ msgid ""
 msgstr ""
 "Project-Id-Version: \n"
 "Report-Msgid-Bugs-To: \n"
-"POT-Creation-Date: 2020-06-24 15:50+0300\n"
+"POT-Creation-Date: 2020-07-03 10:03+0300\n"
 "PO-Revision-Date: 2019-08-14 12:16+0200\n"
 "Last-Translator: Saara Satokangas <saara.satokangas@aalto.fi>\n"
```
Comment (etanttila) [inline locale/fi/LC_MESSAGES/django.po:11]: To my understanding, it's usually unnecessary to update/commit these (as we use git) and their updation easily causes merge-conflicts

File: course/templates/course/_course_cards.html [hunk 6]
```diff
@@ -3,8 +3,7 @@
 
 {% for instance in instances %}
 {% with instance_url=instance|url %}
-<div class="col-sm-6 col-md-4 col-lg-3">
-	<div class="frontpage-course card">
+	<div class="frontpage course-card card">
```
Comment (etanttila) [inline course/templates/course/_course_cards.html:6]: Is this course-card class used for anything? Or what is its intended purpose?

File: assets/css/main.css (hunk 4 above)
Comment (saarasat) [reply assets/css/main.css]: That is a good point! Probably won't change much, but definitely should be logical limits. Those have now been changed at: https://github.com/apluslms/a-plus/pull/608/commits/96a174516db506c94d5314fc0d29acccb0d11a0c

File: locale/fi/LC_MESSAGES/django.po (hunk 5 above)
Comment (saarasat) [reply locale/fi/LC_MESSAGES/django.po:11]: You mean the POT-Creation-Date changes or the translation changes? 

I like to add the changes in translations commit that change them, as I think that the creation-date and who made the changes to translations is important if the state of the branch would be reverted to some specific point (commit) in time. And I also like to ensure that they work before I commit/create a pull request. I guess they could be separated into their own commit at the end of a PR. Though then the branch would not necessarily work after each commit ... hmm.. not sure what is the best way. 

But for example, with my previous pull request, some new translations appeared from other earlier pull requests, since those files had not been committed after their final changes in the related PRs. It looked a bit weird as I hadn't wrote those translations, and at least I try to prevent that from happening.

File: course/templates/course/_course_cards.html (hunk 6 above)
Comment (saarasat) [reply course/templates/course/_course_cards.html]: That has become obsolete, you are correct. I removed that.

**Comment by saarasat** [comment]
> Aalto VDI has IE 11, this change could be checked there. I'm not sure how, but I think the front page should be usable to some extend with that browser.

Here are the results of IE 11, and Edge as well. Both seem to work. 

Screenshot from 2020-07-07 09-29-47

image

File: locale/fi/LC_MESSAGES/django.po (hunk 5 above)
Comment (etanttila) [reply locale/fi/LC_MESSAGES/django.po:11]: I mean the POT-Creation-Date and Last-Translator changes. The translations themselves should definitely be part of the same commit as the things that are changed.
I agree that it's kind of weird that there are new changes that aren't mentioned in the meta-data (and I think it's even more weird that the revision date is before the creation date), but when I've asked @raphendyr about whether to contain those POT-Creation-Dates and Last-Translator changes, he's always instructed to leave them out. He's probably able to give more input on it.

Comment (etanttila) [reply locale/fi/LC_MESSAGES/django.po:11]: I asked him whether to include them or not: "saa, ei oo pakko. pitää keksiä style guideen jotain"

Comment (saarasat) [reply locale/fi/LC_MESSAGES/django.po:11]: Alrighty, that clears it up, thanks :)

**Comment by saarasat** [comment]
@raphendyr @mtthwhggns @etanttila 
Since you all reviewed this PR, would you like confirm if it is okay or is there something you'd like me to add :)?

File: course/templates/course/index.html [hunk 7]
```diff
@@ -31,33 +31,31 @@ <h2 class="panel-heading">{% trans "My courses" %}</h2>
...
+							<img class="card-img-top" src="{{ advert.image }}" alt="" />
+						{% else %}
+							<div class="card-img-top">
+								<i class="glyphicon glyphicon-comment"></i>
+							</div>
+						{% endif %}
+					</div>
+					<div class="card-body">
+						<h3 class="card-title">{{ advert.title }}</h3>
+						<p class="card-text">{{ advert.text|safe }}</p>
```
Comment (jaguarfi) [inline course/templates/course/index.html:50]: Does this behave differently than the other courses? Why?. Btw, how do you mark a course as  `advert`?. I was trying to find this option, but I could not find it.

Comment (saarasat) [reply course/templates/course/index.html:50]: You mean the advert? That whether that behaves differently? Yeah, it was a bit hard to find at first for me too. But there is a sidewide advertisement that you can place from settings.py and it shows like this on the frontpage. So acts like as one of the course-cards:

image

Teachers can't set that for any course basically, it is only for the site-managers to set.

Comment (jaguarfi) [reply course/templates/course/index.html:50]: Great, I did not know about that option :smile: . And yes, I was asking about the advert card.

**Comment by saarasat** [comment]
Commits squashed, and changes tested one more time just in case :)
